### DISCUSSION TITLE
Unable to build standalone app based on google cloud speech code using Makefile

### DISCUSSION THREAD

**Discussion by jiovaibhav** [googleapis/google-cloud-cpp#9817] [Q&A] [answered]
We are trying to build a standalone sample application and calling speech Recognize methods as mentioned in quickstart app.

I have already built google-cloud-cpp repo and its required dependencies for Ubuntu 18.04 on my system as explained in packaging link. I have installed all the produced libraries in /usr/local/lib/
[https://github.com/googleapis/google-cloud-cpp/blob/main/doc/packaging.md](url)

Post building I utilised these libraries in my Makefile to build cpp code but it fails with "undefined references" errors. To resolve the errors,  I have included all the libraries inside my vcpkg :

VCLIBS = $(wildcard /root/vcpkg/packages/*/lib/*.a)
VCINLIBS = $(wildcard /root/vcpkg/installed/x64-linux/lib/*.a)

I included all libraries downloaded/installed by vcpkg and referred them in my Makefile which executes by code successfully. I am also able to get results from Recognize and AsyncStreamingRecognize methods from my cpp sample code.

I sharing my Makefile which includes all the dependencies I have tried to build the code. I have commented out the references I have used after building google-cloud-cpp.

[sample-github.txt](https://github.com/googleapis/google-cloud-cpp/files/9540327/sample-github.txt)
[github-Makefile.txt](https://github.com/googleapis/google-cloud-cpp/files/9540329/github-Makefile.txt)

Although I followed the steps as per the website, but doubt that I am doing this correctly. I might have missed some steps or have installed wrong versions of dependencies but I am not sure about this.

Ideally I should have only used libraries and header files from /usr/local/ folder. 

Can someone please look into my Makefile and suggest the corrections or guide me to the right path ?

**Comment by coryan** [comment, marked as answer]
> I have already built google-cloud-cpp repo and its required dependencies for Ubuntu 18.04 on my system as explained in packaging link. I have installed all the produced libraries in /usr/local/lib/ [https://github.com/googleapis/google-cloud-cpp/blob/main/doc/packaging.md](url)

Hmmm.

> 
> Post building I utilised these libraries in my Makefile to build cpp code but it fails with "undefined references" errors. To resolve the errors, I have included all the libraries inside my vcpkg :

I am confused. Why are you trying to use the packaging instructions **and** use vcpkg?  If you use vcpkg you don't need to install `google-cloud-cpp` in `/usr/local/`, vcpkg will install it for you in its preferred install location.  If you install dependencies in both locations you will need to ensure both use compatible compilers and compiler flags.  That seems very difficult.

> I included all libraries downloaded/installed by vcpkg and referred them in my Makefile which executes by code successfully. I am also able to get results from Recognize and AsyncStreamingRecognize methods from my cpp sample code.

It sounds like you are asking how to simplify all of this, as the approach you followed "works", but it is obviously very complicated and fraught with problems.  If I misunderstand your request please let me know.  The first thing I would do is pick one approach for dependency management and get that approach to work.  Since you are willing to use `vcpkg`, then I would recommend:

1) Use `vcpkg` to manage all your dependencies.
2) Do **not** install any dependencies or `google-cloud-cpp` in `/usr/local`

You can do the opposite if that works better for you.  I will avoid making further recommendations until I understand what is your next step.

> 
> I sharing my Makefile which includes all the dependencies I have tried to build the code. I have commented out the references I have used after building google-cloud-cpp.
> 
> [sample-github.txt](https://github.com/googleapis/google-cloud-cpp/files/9540327/sample-github.txt) [github-Makefile.txt](https://github.com/googleapis/google-cloud-cpp/files/9540329/github-Makefile.txt)
> 
> Although I followed the steps as per the website, but doubt that I am doing this correctly. I might have missed some steps or have installed wrong versions of dependencies but I am not sure about this.
> 
> Ideally I should have only used libraries and header files from /usr/local/ folder.
> 
> Can someone please look into my Makefile and suggest the corrections or guide me to the right path ?

I cannot give you detailed instructions until I understand which approach (vcpkg vs. installing in /usr/local) you are trying to use.

Meanwhile I can provide some suggestions:

Your `github-Makefile.txt` includes:

```Makefile
VCLIBS = $(wildcard /root/vcpkg/packages/*/lib/*.a)
```

If I was writing a `Makefile` I would not use `$(wildcard)` to find the `.a` files.  These files need to be pass to the linker in their (topological) order of depedendencies, i.e., if `libA.a` depends on `libB.a` then `libA.a` (or `-lA`) must be first in the linker command line.

Your `github-Makefile.txt` also includes:

```Makefile

LIBS                       = -L/root/vcpkg/packages/google-cloud-cpp_x64-linux/lib \
                            -lgoogle_cloud_cpp_api_context_protos	 -lgoogle_cloud_cpp_api_quota_protos     -lgoogle_cloud_cpp_speech			   -lgoogle_cloud_cpp_type_phone_number_protos \
                            -lgoogle_cloud_cpp_speech_protos		   -lgoogle_cloud_cpp_type_postal_address_protos \
                            -lgoogle_cloud_cpp_grpc_utils \```
...
...
```

If I was writing a Makefile I would not curate the list of libraries manually.

The `google/cloud/speech/Makefile` contains what I would recommend instead of those two options:

https://github.com/googleapis/google-cloud-cpp/blob/a7d1919a925b013c65cf190f291250e33a766d23/google/cloud/speech/quickstart/Makefile#L29-L32

That *should* work regardless of whether you use `vcpkg` *or* you install in `/usr/local/`, but I just filed https://github.com/microsoft/vcpkg/issues/26751 which means the `vcpkg` approach is currently broken.

**Reply by jiovaibhav** [reply]
> I am confused. Why are you trying to use the packaging instructions and use vcpkg?

The aim was to build google-cloud-cpp and use the dependencies to build and run my code. But that did not happen, so I switched to vcpkg. I was not sure how to use it in my Makefile based project, so imported all the libraries I found. I know it is not a good practice to include use wildcards and add libraries, but it did work for me. Guess I was lucky there !!

That should work regardless of whether you use vcpkg or you install in /usr/local/, but I just filed https://github.com/microsoft/vcpkg/issues/26751 which means the vcpkg approach is currently broken.

I took the latest version of vcpkg this morning and found the same problem still persists.

```
/root/vcpkg/installed/x64-linux/lib/pkgconfig/../../lib/libupb.a(decode.c.o): In function decode_msg':
decode.c:(.text+0xb31): undefined reference to utf8_naive' decode.c:(.text+0xbf1): undefined reference to utf8_naive'
decode.c:(.text+0xf71): undefined reference to utf8_naive' decode.c:(.text+0x1049): undefined reference to utf8_naive'`
```
Can you please confirm if your PR for vcpkg has fixed the issue ?

**Reply by coryan** [reply]
> Can you please confirm if your PR for vcpkg has fixed the issue ?

Yes. That is exactly the error I got, and the PR fixes it.

If you don't want to wait for that PR to get merged, you can tweak the `Makefile` as follows:

```diff
diff --git a/Makefile b/Makefile
index bdabe82..4be7616 100644
--- a/Makefile
+++ b/Makefile
@@ -29,7 +29,7 @@ all: $(BIN)/quickstart
 CLIENT_MODULE     := google_cloud_cpp_speech
 CLIENT_CXXFLAGS   := $(shell pkg-config $(CLIENT_MODULE) --cflags)
 CLIENT_CXXLDFLAGS := $(shell pkg-config $(CLIENT_MODULE) --libs-only-L)
-CLIENT_LIBS       := $(shell pkg-config $(CLIENT_MODULE) --libs-only-l)
+CLIENT_LIBS       := $(shell pkg-config $(CLIENT_MODULE) --libs-only-l) -lupb_fastdecode -lupb_utf8_range
 
 $(BIN)/quickstart: quickstart.cc
        $(CXXLD) $(CXXFLAGS) $(CLIENT_CXXFLAGS) $(CLIENT_CXXLDFLAGS) -o $@ $^ $(CLIENT_LIBS)
```

Alternatively, you can apply my PR (https://github.com/microsoft/vcpkg/pull/26752) to your copy of `vcpkg`.  Or, if you are comfortable with `vcpkg` overlays, you could create an overlay for `grpc` that includes the changes in my PR.

Let me know if you want to pursue getting the libraries manually installed in `/usr/local`.  I am not sure what failed so I cannot offer any sensible advice.

**Reply by jiovaibhav** [reply]
```
diff --git a/Makefile b/Makefile
index bdabe82..4be7616 100644
--- a/Makefile
+++ b/Makefile
@@ -29,7 +29,7 @@ all: $(BIN)/quickstart
 CLIENT_MODULE     := google_cloud_cpp_speech
 CLIENT_CXXFLAGS   := $(shell pkg-config $(CLIENT_MODULE) --cflags)
 CLIENT_CXXLDFLAGS := $(shell pkg-config $(CLIENT_MODULE) --libs-only-L)
-CLIENT_LIBS       := $(shell pkg-config $(CLIENT_MODULE) --libs-only-l)
+CLIENT_LIBS       := $(shell pkg-config $(CLIENT_MODULE) --libs-only-l) -lupb_fastdecode -lupb_utf8_range
 
 $(BIN)/quickstart: quickstart.cc
        $(CXXLD) $(CXXFLAGS) $(CLIENT_CXXFLAGS) $(CLIENT_CXXLDFLAGS) -o $@ $^ $(CLIENT_LIBS)
```


Thanks @coryan , adding these libs solves the issue by vcpkg
